test/python/circuit/test_hamiltonian_gate.py

```diff
@@ -0,0 +1,201 @@
+        qc2 = qc2.bind_parameters({theta: np.pi / 2})
+        # test of qasm output
+        self.log.info(qc2.qasm())
+        # test of text drawer
+        self.log.info(qc2)
+        dag = circuit_to_dag(qc2)
+        nodes = dag.two_qubit_ops()
+        self.assertEqual(len(nodes), 1)
+        dnode = nodes[0]
+        self.assertIsInstance(dnode.op, HamiltonianGate)
+        for qubit in dnode.qargs:
+            self.assertIn(qubit.index, [0, 1])
+        # Equality based on Pauli exponential identity
+        np.testing.assert_array_almost_equal(dnode.op.to_matrix(), 1j*matrix.data)
+        qc3 = dag_to_circuit(dag)
+        self.assertEqual(qc2, qc3)
+
+    def test_3q_hamiltonian(self):
+        """test 3 qubit hamiltonian on non-consecutive bits"""
+        qr = QuantumRegister(4)
+        qc = QuantumCircuit(qr)
+        qc.x(qr[0])
+        matrix = Operator.from_label('XZY')
+        theta = Parameter('theta')
+        uni3q = HamiltonianGate(matrix, theta)
+        qc.append(uni3q, [qr[0], qr[1], qr[3]])
+        qc.cx(qr[3], qr[2])
+        # test of text drawer
+        self.log.info(qc)
+        qc = qc.bind_parameters({theta: np.pi / 2})
+        dag = circuit_to_dag(qc)
+        nodes = dag.multi_qubit_ops()
+        self.assertEqual(len(nodes), 1)
+        dnode = nodes[0]
+        self.assertIsInstance(dnode.op, HamiltonianGate)
+        for qubit in dnode.qargs:
+            self.assertIn(qubit.index, [0, 1, 3])
+        np.testing.assert_almost_equal(dnode.op.to_matrix(), 1j*matrix.data)
+
+    def test_qobj_with_hamiltonian(self):
+        """test qobj output with hamiltonian"""
+        qr = QuantumRegister(4)
+        qc = QuantumCircuit(qr)
+        qc.rx(np.pi / 4, qr[0])
+        matrix = Operator.from_label('XIZ')
+        theta = Parameter('theta')
+        uni = HamiltonianGate(matrix, theta, label='XIZ')
+        qc.append(uni, [qr[0], qr[1], qr[3]])
+        qc.cx(qr[3], qr[2])
+        qc = qc.bind_parameters({theta: np.pi / 2})
+        qobj = qiskit.compiler.assemble(qc)
+        instr = qobj.experiments[0].instructions[1]
+        self.assertEqual(instr.name, 'hamiltonian')
+        # Also test label
+        self.assertEqual(instr.label, 'XIZ')
+        np.testing.assert_array_almost_equal(np.array(instr.params[0]).astype(np.complex64),
+                                             matrix.data)
+        # check conversion to dict
+        qobj_dict = qobj.to_dict(validate=True)
+
+        class NumpyEncoder(json.JSONEncoder):
+            """Class for encoding json str with complex and np arrays."""
+            def default(self, obj):
+                if isinstance(obj, np.ndarray):
+                    return obj.tolist()
+                if isinstance(obj, complex):
+                    return (obj.real, obj.imag)
+                return json.JSONEncoder.default(self, obj)
+
+        # check json serialization
+        self.assertTrue(isinstance(json.dumps(qobj_dict, cls=NumpyEncoder),
```

Comment:
The encoder is defined as part of the test and I dont see the point of testing it. Either the encoder should be part of the code base or the test is testing the test...

Reply:
This is basically what the ibmq provider uses internally now that qobj is marshmallow free: https://github.com/Qiskit/qiskit-ibmq-provider/blob/2c4d04dc91ddd495643a69c82dd92434bfc21603/qiskit/providers/ibmq/utils/json_encoder.py so I think it's fine to test this if we need to. It's something we do in the qobj tests for example. But I agree in this case testing that the qobj object post json serialization doesn't provide too much value. We're not actually testing anything outside of what is already covered by qobj testing and ibmq provider testing. But in those cases we test that the object is preserved through serialization (by reloading it) just asserting it's a string doesn't seem to useful. That being said I'm not sure if it's really a blocker here, its something we can always cleanup in a follow up pr.

Reply:
Removed!